File: packages/java/src/main/java/ai/alumnium/tool/UploadTool.java
package ai.alumnium.tool;

import ai.alumnium.driver.BaseDriver;
import ai.alumnium.tool.annotation.ToolDescription;
import ai.alumnium.tool.annotation.ToolField;
import java.util.List;

@ToolDescription(
    "Upload one or more files using a button that opens a file chooser. This tool automatically"
        + " clicks the button, DO NOT use ClickTool for that.")
public record UploadTool(
    @ToolField(description = "Element identifier (ID)") int id,
    @ToolField(
            description =
                "Absolute file path(s) to upload. Can be a single path or multiple"
                    + " paths for multi-file upload.")
        List<String> paths)
    implements BaseTool {
  @Override
  public void invoke(BaseDriver driver) {
    driver.upload(id, paths);
  }
}

File: packages/python/src/alumnium/tools/upload_tool.py
from re import sub

from pydantic import Field

from alumnium.drivers.base_driver import BaseDriver

from .base_tool import BaseTool


class UploadTool(BaseTool):
    __doc__ = (
        "Upload one or more files using a button that opens a file chooser. "
        "This tool automatically clicks the button, DO NOT use ClickTool for that."
    )

    id: int = Field(description="Element identifier (ID)")
    paths: list[str] = Field(
        description="Absolute file path(s) to upload. Can be a single path or multiple paths for multi-file upload."
    )

    def invoke(self, driver: BaseDriver):
        driver.upload(self.id, self._normalize_paths(self.paths))  # type: ignore[reportAttributeAccessIssue]  # ty:ignore[unresolved-attribute]

    def _normalize_paths(self, paths: list[str]) -> list[str]:
        # Planner often attempts to "escape" file paths by adding backslashes.
        # It also often surrounds paths with quotes.
        normalized = []
        for path in paths:
            normalized_path = sub(r"\\+/", "/", path).strip('"').strip("'").strip()
            normalized.append(normalized_path)

        return normalized
